go/libraries/doltcore/sqle/tables.go

```diff
@@ -498,57 +497,292 @@ func (t *AlterableDoltTable) CreateIndex(ctx *sql.Context, indexName string, usi
+			return fmt.Errorf("table `%s` does not have column `%s`", t.name, col)
+		}
+		tblCols[i] = tableCol
+		colTags[i] = tableCol.Tag
+		sqlColNames[i] = sql.IndexColumn{
+			Name:   tableCol.Name,
+			Length: 0,
+		}
+	}
+
+	refColTags := make([]uint64, len(referencedColumns))
+	sqlRefColNames := make([]sql.IndexColumn, len(referencedColumns))
+	for i, col := range referencedColumns {
+		tableCol, ok := refTableSch.GetAllCols().GetByNameCaseInsensitive(col)
+		if !ok {
+			return fmt.Errorf("table `%s` does not have column `%s`", refTableName, col)
+		}
+		refColTags[i] = tableCol.Tag
+		sqlRefColNames[i] = sql.IndexColumn{
+			Name:   tableCol.Name,
+			Length: 0,
+		}
+		if !tblCols[i].TypeInfo.Equals(tableCol.TypeInfo) {
+			//TODO: make this more descriptive about our lack of support of different types for fks, figure it out later
+			return fmt.Errorf("column type mismatch on `%s` and `%s`", columns[i], tableCol.Name)
+		}
+		sqlparserType := tableCol.TypeInfo.ToSqlType().Type()
+		if sqlparserType == sqltypes.Blob || sqlparserType == sqltypes.Text {
+			return fmt.Errorf("TEXT/BLOB are not valid types for foreign keys")
+		}
+	}
+
+	onUpdateRefOp, err := t.parseFkReferenceOption(onUpdate)
+	if err != nil {
+		return err
+	}
+	onDeleteRefOp, err := t.parseFkReferenceOption(onDelete)
 	if err != nil {
 		return err
 	}
 
-	root, err := t.db.GetRoot(ctx)
+	newTable, _, tableIndex, err := t.createIndex(ctx, t.table, t.sch, true, "", sql.IndexUsing_Default, sql.IndexConstraint_None, sqlColNames, "")
+	if err != nil {
+		return err
+	}
+	newRefTable, _, refTableIndex, err := t.createIndex(ctx, refTable, refTableSch, true, "", sql.IndexUsing_Default, sql.IndexConstraint_None, sqlRefColNames, "")
 	if err != nil {
 		return err
 	}
+
 	newRoot, err := root.PutTable(ctx, t.name, newTable)
 	if err != nil {
 		return err
 	}
-	return t.db.SetRoot(ctx, newRoot)
-}
+	newRoot, err = newRoot.PutTable(ctx, refTableName, newRefTable)
+	if err != nil {
+		return err
+	}
 
-// RenameIndex implements sql.IndexAlterableTable
-func (t *AlterableDoltTable) RenameIndex(ctx *sql.Context, fromIndexName string, toIndexName string) error {
-	// RenameIndex will error if there is a name collision or an index does not exist
-	_, err := t.sch.Indexes().RenameIndex(fromIndexName, toIndexName)
+	foreignKeyCollection, err := newRoot.GetForeignKeyCollection(ctx)
 	if err != nil {
 		return err
 	}
-	newTable, err := t.table.UpdateSchema(ctx, t.sch)
+	_, err = foreignKeyCollection.AddForeignKey(fkName, t.name, tableIndex.Name(), colTags, refTableName, refTableIndex.Name(), refColTags, onDeleteRefOp, onUpdateRefOp)
+	if err != nil {
+		return err
+	}
+	newRoot, err = newRoot.PutForeignKeyCollection(ctx, foreignKeyCollection)
 	if err != nil {
 		return err
 	}
 
+	err = t.db.SetRoot(ctx, newRoot)
+	if err != nil {
+		return err
+	}
+	return t.updateFromRoot(ctx, newRoot)
+}
+
+func (t *AlterableDoltTable) DropForeignKey(ctx *sql.Context, fkName string) error {
```

Comment:
Need golang tests of these new methods ASAP (can be engine level, but should examine the root object for the presence of the correct constraints and indexes)